### PR TITLE
Crowdloan Dissolve Requires All Funds Refunded

### PR DESCRIPTION
**Pull request by shawntabrizi**
Waiting for: https://github.com/paritytech/substrate/pull/8478

This PR changes the logic for how a crowdloan is dissolved.

Before, there was a retirement state, where users had some limited window to withdraw their funds.

This PR removes this logic and instead requires that all funds are returned to users before the crowdloan can be dissolved.

It also introduces a nice helper function which will refund as many accounts as possible at once, rather than needing to call each refund one by one.

### CODE REVIEW

File: runtime/common/src/crowdloan.rs [hunk 1]
```diff
@@ -59,17 +59,18 @@
...
 	},
+	Identity, StorageHasher, ReversibleStorageHasher,
 	pallet_prelude::{Weight, DispatchResultWithPostInfo},
 };
 use frame_system::{ensure_signed, ensure_root};
 use sp_runtime::{
 	ModuleId, DispatchResult, RuntimeDebug, MultiSignature, MultiSigner,
 	traits::{
-		AccountIdConversion, Hash, Saturating, Zero, One, CheckedAdd, Bounded, Verify, IdentifyAccount,
+		AccountIdConversion, Saturating, Zero, One, CheckedAdd, Verify, IdentifyAccount, Hash,
```
Comment (shawntabrizi) [inline runtime/common/src/crowdloan.rs:73]:
```suggestion
		AccountIdConversion, Hash, Saturating, Zero, One, CheckedAdd, Verify, IdentifyAccount,
```

File: runtime/common/src/crowdloan.rs [hunk 2]
```diff
@@ -160,8 +147,8 @@ pub enum LastContribution<BlockNumber> {
 #[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug)]
 #[codec(dumb_trait_bound)]
 pub struct FundInfo<AccountId, Balance, BlockNumber, LeasePeriod> {
-	/// True if the fund is being retired. This can only be set once and only when the current
-	/// lease period is greater than the `last_period`.
+	/// TODO: retiring is currently unused, but we want to leave it in for Rococo compatibility.
```
Comment (kianenigma) [inline runtime/common/src/crowdloan.rs:150]: linked issue would be nice.

**Review by Az111z — APPROVED** [review]

File: runtime/common/src/crowdloan.rs [hunk 3]
```diff
@@ -465,52 +448,79 @@ decl_module! {
...
-					// Remove all other balance from the account into orphaned funds.
-					let account = Self::fund_account_id(index);
-					let (imbalance, _) = CurrencyOf::<T>::slash(&account, BalanceOf::<T>::max_value());
-					T::OrphanedFunds::on_unbalanced(imbalance);
+			// Save the changes.
+			Funds::<T>::insert(index, &fund);
 
-					Funds::<T>::remove(index);
+			if all_refunded {
+				Self::deposit_event(RawEvent::AllRefunded(index));
```
Comment (kianenigma) [inline runtime/common/src/crowdloan.rs:492]: and if I call this dispatch again after `AllRefunded`?

File: runtime/common/src/crowdloan.rs [hunk 4]
```diff
@@ -465,52 +448,79 @@ decl_module! {
...
+				if refund_count >= T::RemoveKeysLimit::get() {
+					// Not everyone was able to be refunded this time around.
+					all_refunded = false;
+					break;
+				}
+				CurrencyOf::<T>::transfer(&fund_account, &who, balance, AllowDeath)?;
+				Self::contribution_kill(fund.trie_index, &who);
+				fund.raised = fund.raised.saturating_sub(balance);
+				refund_count += 1;
+			}
```
Comment (kianenigma) [inline runtime/common/src/crowdloan.rs:486]:
```suggestion
			let contributions = Self::contribution_iterator(fund.trie_index);
			let refunded = contributions.into_iter().take(T::RemoveKeysLimit::get()).fold(Zero::zero(), |refunded, (who, (balance, _))| {
				CurrencyOf::<T>::transfer(&fund_account, &who, balance, AllowDeath)?;
				Self::contribution_kill(fund.trie_index, &who);
				fund.raised = fund.raised.saturating_sub(balance);
				refunded += 1 
			});
			
			if refunded == T::RemoveKeysLimit::get() {
			  // everyone was refunded
			} else {
			  // some left 
			}
```

I find this simpler but just a suggestion

File: runtime/common/src/crowdloan.rs [hunk 5]
```diff
@@ -465,52 +448,79 @@ decl_module! {
...
+
+			// Only allow dissolution when the raised funds goes to zero,
+			// and the caller is the fund creator or we are past the end date.
+			let can_dissolve = fund.raised.is_zero()
+				&& (who == fund.depositor || now >= fund.end);
+			ensure!(can_dissolve, Error::<T>::NotReadyToDissolve);
+
+			// Assuming state is not corrupted, the child trie should already be cleaned up
+			// an all funds in the crowdloan account have been returned. If not, governance
+			// can take care of that.
```
Comment (kianenigma) [inline runtime/common/src/crowdloan.rs:518]:
```suggestion
			// Assuming state is not corrupted, the child trie should already be cleaned up
			// an all funds in the crowdloan account have been returned. If not, governance
			// can take care of that.
			debug_assert!(Self::contribution_iterator(fund.trie_index).count().is_zero());
```

**Review by kianenigma — APPROVED** [review]
A few suggestions, but looks solid overall.

File: runtime/common/src/crowdloan.rs [hunk 6]
```diff
@@ -160,8 +147,8 @@ pub enum LastContribution<BlockNumber> {
 #[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug)]
 #[codec(dumb_trait_bound)]
 pub struct FundInfo<AccountId, Balance, BlockNumber, LeasePeriod> {
-	/// True if the fund is being retired. This can only be set once and only when the current
-	/// lease period is greater than the `last_period`.
+	/// TODO: retiring is currently unused, but we want to leave it in for Rococo compatibility.
+	/// Just need to delete this first item.
 	retiring: bool,
```
Comment (shawntabrizi) [inline runtime/common/src/crowdloan.rs:152]:
```suggestion
```

File: runtime/common/src/crowdloan.rs (hunk 3 above)
Comment (shawntabrizi) [reply runtime/common/src/crowdloan.rs:489]: the same event is fired again?

dont really see the need to do additional checks and emit an error

File: runtime/common/src/crowdloan.rs [hunk 7]
```diff
@@ -32,44 +32,36 @@
...
 //!
-//! Contributors may get a refund of their contributions from retired funds. After a period (`RetirementPeriod`)
-//! the fund may be dissolved entirely. At this point any non-refunded contributions are considered
-//! `orphaned` and are disposed of through the `OrphanedFunds` handler (which may e.g. place them
-//! into the treasury).
+//! Contributors will get a refund of their contributions from completed funds before the crowdloan
+//! can be dissolved.
 //!
-//! Funds may accept contributions at any point before their success or retirement. When a parachain
+//! Funds may accept contributions at any point before their success. When a parachain
```
Comment (shawntabrizi) [inline runtime/common/src/crowdloan.rs:44]:
```suggestion
//! Funds may accept contributions at any point before their success or end. When a parachain
```

File: runtime/common/src/crowdloan.rs [hunk 8]
```diff
@@ -465,52 +444,80 @@ decl_module! {
...
-					Ok(Some(T::WeightInfo::dissolve(num_removed)).into())
-				}
-			}
+			let fund = Self::funds(index).ok_or(Error::<T>::InvalidParaId)?;
+			let now = frame_system::Pallet::<T>::block_number();
+
+			// Only allow dissolution when the raised funds goes to zero,
+			// and the caller is the fund creator or we are past the end date.
+			let can_dissolve = fund.raised.is_zero()
+				&& (who == fund.depositor || now >= fund.end);
```
Comment (gavofyork) [inline runtime/common/src/crowdloan.rs:509]:
```suggestion
			let permitted = who == fund.depositor || now >= fund.end;
			let can_dissolve = permitted && fund.raised.is_zero();
```

File: runtime/common/src/crowdloan.rs (hunk 3 above)
Comment (kianenigma) [reply runtime/common/src/crowdloan.rs:488]: So no harmful side-effects.

**Review by gavofyork — APPROVED** [review]

**Comment by shawntabrizi** [comment]
bot merge

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**